File: native/python/tests/real_conformance.py
import argparse
import asyncio
import json
import threading
from pathlib import Path

from rti_demo_ui import DemoUiApp
from rti_demo_ui_native import run_native


EXPECTED_CHECKS = {
    "snapshot",
    "sse",
    "dynamic_import",
    "runtime3d_import",
    "module_worker",
    "theme_asset",
    "persistent_storage",
    "canvas",
    "webgl",
    "keyboard_focus",
    "resize_observation",
    "navigation_policy",
    "command_origin",
}


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--application-id", required=True)
    parser.add_argument("--expected", required=True)
    parser.add_argument("--write", required=True)
    parser.add_argument("--static-root", required=True, type=Path)
    args = parser.parse_args()

    report_event = threading.Event()
    report = {}
    app = DemoUiApp("Native Phase 2 conformance", static_root=args.static_root)
    app.set_data(
        {
            "native_storage_key": "phase2-production",
            "native_storage_expected": args.expected,
            "native_storage_write": args.write,
        }
    )

    def record_report(payload):
        report.update(payload)
        report_event.set()
        return {"recorded": True}

    app.register_command("spike-report", {"type": "object"}, record_report)
    app.register_command(
        "spike-origin",
        {"type": "object"},
        lambda payload: {"origin": payload["origin"]},
    )

    async def wait_for_report(_app):
        if not await asyncio.to_thread(report_event.wait, 20):
            raise TimeoutError("production conformance report timed out")

    run_native(
        app,
        application_id=args.application_id,
        async_main=wait_for_report,
    )

    results = report.get("results", {})
    missing = EXPECTED_CHECKS.difference(results)
    failed = {
        name: result
        for name, result in results.items()
        if not result.get("passed", False)
    }
    if missing or failed:
        raise RuntimeError(
            f"production conformance failed: missing={sorted(missing)} failed={failed}"
        )
    print(json.dumps(report, sort_keys=True))


if __name__ == "__main__":
    main()

File: native/cpp/tests/real_conformance.cpp
#include <rti_demo_ui_native/native_webview.hpp>

#include <chrono>
#include <condition_variable>
#include <exception>
#include <iostream>
#include <mutex>
#include <set>
#include <string>
#include <thread>

namespace {

using rti::demo::ui::CommandSchema;
using rti::demo::ui::DemoUiApp;
using rti::demo::ui::Json;

const std::set<std::string> expected_checks{
    "snapshot",          "sse",             "dynamic_import",
    "runtime3d_import",  "module_worker",   "theme_asset",
    "persistent_storage", "canvas",          "webgl",
    "keyboard_focus",    "resize_observation", "navigation_policy",
    "command_origin",
};

}  // namespace

int main(int argc, char** argv) {
    if (argc != 3) {
        std::cerr << "usage: real_conformance STORAGE_EXPECTED STORAGE_WRITE\n";
        return 2;
    }

    DemoUiApp app("Native Phase 2 conformance", 0, "127.0.0.1",
                  NATIVE_CONFORMANCE_STATIC_ROOT);
    app.set_data(Json{{"native_storage_key", "phase2-production"},
                      {"native_storage_expected", argv[1]},
                      {"native_storage_write", argv[2]}});

    std::mutex mutex;
    std::condition_variable report_cv;
    Json report;
    bool report_received = false;
    app.register_command(
        "spike-report", CommandSchema(Json{{"type", "object"}}),
        [&](const Json& payload) {
            {
                std::lock_guard<std::mutex> guard(mutex);
                report = payload;
                report_received = true;
            }
            report_cv.notify_all();
            return Json{{"recorded", true}};
        });
    app.register_command(
        "spike-origin", CommandSchema(Json{{"type", "object"}}),
        [](const Json& payload) { return Json{{"origin", payload["origin"]}}; });

    std::thread closer([&]() {
        std::unique_lock<std::mutex> lock(mutex);
        report_cv.wait_for(lock, std::chrono::seconds(20),
                           [&]() { return report_received; });
        lock.unlock();
        app.stop();
    });

    try {
        rti::demo::ui::native::run(app);
    } catch (const std::exception& error) {
        app.stop();
        closer.join();
        std::cerr << error.what() << '\n';
        return 1;
    }
    closer.join();

    if (!report_received || !report.contains("results")) {
        std::cerr << "production conformance report timed out\n";
        return 1;
    }
    const auto& results = report["results"];
    for (const auto& name : expected_checks) {
        if (!results.contains(name) || !results[name].value("passed", false)) {
            std::cerr << "failed conformance check: " << name << '\n';
            return 1;
        }
    }
    std::cout << report.dump() << '\n';
    return 0;
}
